src/category_theory/closed/cartesian.lean

```diff
@@ -0,0 +1,405 @@
+
+Show that exponential forms a difunctor and define the exponential comparison morphisms.
+
+## TODO
+Some of the results here are true more generally for closed objects and
+for closed monoidal categories, and these could be generalised.
+-/
+universes v u u₂
+
+namespace category_theory
+
+open category_theory category_theory.category category_theory.limits
+
+local attribute [instance] monoidal_of_has_finite_products
+
+/--
+An object `X` is *exponentiable* if `(X × -)` is a left adjoint.
+We define this as being `closed` in the cartesian monoidal structure.
+-/
+abbreviation exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] (X : C) :=
+closed X
+
+/--
+If `X` and `Y` are exponentiable then `X ⨯ Y` is.
+This isn't an instance because it's not usually how we want to construct exponentials, we'll usually
+prove all objects are exponential uniformly.
+-/
+def binary_product_exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] {X Y : C}
+  (hX : exponentiable X) (hY : exponentiable Y) : exponentiable (X ⨯ Y) :=
+{ is_adj :=
+  begin
+    haveI := hX.is_adj,
+    haveI := hY.is_adj,
+    exact adjunction.left_adjoint_of_nat_iso (monoidal_category.tensor_left_tensor _ _).symm
+  end }
+
+/--
+The terminal object is always exponentiable.
+This isn't an instance because most of the time we'll prove cartesian closed for all objects
+at once, rather than just for this one.
+-/
+def terminal_exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] : exponentiable ⊤_C :=
+{ is_adj :=
+  { right := 𝟭 C,
+    adj := adjunction.mk_of_hom_equiv
+    { hom_equiv := λ X _, have unitor : _, from prod.left_unitor X,
+        ⟨λ a, unitor.inv ≫ a, λ a, unitor.hom ≫ a, by tidy, by tidy⟩ } } }
+
+/--
+A category `C` is cartesian closed if it has finite products and every object is exponentiable.
+We define this as `monoidal_closed` with respect to the cartesian monoidal structure.
+-/
+abbreviation cartesian_closed (C : Type u) [category.{v} C] [has_finite_products.{v} C] :=
+monoidal_closed C
+
+variables {C : Type u} [category.{v} C] (A B : C) {X X' Y Y' Z : C}
+
+section exp
+variables [has_finite_products.{v} C] [exponentiable A]
+
+/-- This is (-)^A. -/
+def exp.functor : C ⥤ C :=
+(@closed.is_adj _ _ _ A _).right
+
+/-- The adjunction between A ⨯ - and (-)^A. -/
+def exp.adjunction : prod_functor.obj A ⊣ exp.functor A :=
+closed.is_adj.adj
+
+/-- The evaluation natural transformation. -/
+def ev.nat_trans : exp.functor A ⋙ prod_functor.obj A ⟶ 𝟭 C :=
+closed.is_adj.adj.counit
+
+/-- The coevaluation natural transformation. -/
+def coev.nat_trans : 𝟭 C ⟶ prod_functor.obj A ⋙ exp.functor A :=
+closed.is_adj.adj.unit
+
+/-- `B ^ A` or `A ⟹ B` -/
+def exp : C := (exp.functor A).obj B
+
+infixl ` ⟹ `:20 := exp
+infixr `^^`:30 := pow
+
+/-- Postcompose an internal hom with an external hom. -/
+def post (f : X ⟶ Y) : A⟹X ⟶ A⟹Y :=
+(exp.functor A).map f
```

Comment:
I'm not sure if this is a good idea. The alternative is to work with the exponential functor directly (calling it `exp` rather than `exp.functor`) and so these would just be standard functorial lemmas and wouldn't need naming, as opposed to the `post` indirection. This would be more in line with `prod_functor` and `functor.const`, for instance. Another option is to get to the difunctor quickly, and use that (the downside here is that it requires that every object is exponentiable). Feedback wanted!

Reply:
The alternative sounds good to me?